src/commands/aaleaderboard.rs

```diff
@@ -0,0 +1,166 @@
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+use std::{collections::HashMap, time::{Duration, SystemTime, UNIX_EPOCH}};
+
+macro_rules! return_if_err {
+    ($e:expr) => {
+        match $e {
+            Err(err) => return err,
+            Ok(v) => v,
+        }
+    };
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct AAPlayer {
+    name: String,
+    rank: u32,
+    #[serde(rename = "runTime")]
+    igt: String,
+    // #[serde(rename = "dateAccomplished")]
+    // date: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct APIResponse {
+    players: HashMap<String, AAPlayer>,
+    leaderboard: Vec<String>,
+    #[serde(rename = "lastUpdated")]
+    last_updated: String,
+}
+
+#[derive(Default)]
+pub struct AALeaderboard {
+    data: Option<APIResponse>,
+    next_fetch: Option<SystemTime>,
+}
+
+impl std::fmt::Display for AAPlayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let name = &self.name;
+        let igt = &self.igt;
+        write!(f, "{} ({})", name, igt)
+    }
+}
+
+impl AALeaderboard {
+    pub fn new() -> Self {
+        Self { ..Default::default() }
+    }
+
+    pub async fn fetch_if_required(&mut self) {
+        let now = SystemTime::now();
+
+        if self.data.is_none() || self.next_fetch.map(|t| t <= now).unwrap_or(true) {
+            self.fetch().await
+        }
+    }
+
+    pub async fn fetch(&mut self) {
+        const URL: &str = "https://totorewa.github.io/aa-leaderboard/1-16-1.json";
+        const REFRESH_AFTER: Duration = Duration::from_secs(60 * 60);
+        const RETRY_AFTER: Duration = Duration::from_secs(60 * 5);
+
+        let res = match reqwest::get(URL).await {
+            Ok(res) => res,
+            Err(err) => {
+                println!("{}", err);
+                return
+            }
+        };
+        match res.json::<APIResponse>().await {
+            Ok(lb) => {
+                self.data = Some(lb);
+                self.next_fetch = Some(SystemTime::now() + REFRESH_AFTER);
+            },
+            Err(err) => {
+                println!("{}", err);
+                self.data = None;
+                self.next_fetch = Some(SystemTime::now() + RETRY_AFTER);
+            }
+        }
+    }
+
+    pub fn is_loaded(&self) -> bool {
+        self.next_fetch.is_some() && self.data.is_some()
+    }
+
+    pub fn info_at_rank(&self, rank: u32) -> String {
+        if rank < 1 || rank > 100 {
+            return "Sorry, I only keep track of the top 100 players.".to_string();
+        }
+
+        let lb = return_if_err!(self.get_data());
+        let normalized_name = &lb.leaderboard[rank as usize - 1];
+        AALeaderboard::info_for_normalized_name(lb, normalized_name)
```

Comment:
Array access does not check the index and assumes that the array is of length 100. While this should be the case, it's better to be safe anyway and use `.get(rank - 1)` instead.

Reply:
Fixed